fix all unit failed in vaapi directory.

xuguangxin opened the pull request:
There are vaapi display compatible  issues. This will fix https://github.com/01org/libyami/issues/648

xuguangxin committed:
vaapi: any display type should compatiable with auto

this will fix unit test break by previous commit

xuguangxin committed:
vaapi: fix x11/drm compatible problem

previous code falsely think all X11 display is compatible with drm display.
it’s not true.they are only compatible when handle is invalid. Else we need
find right drm display for user

xuguangxin committed:
unittest: NATIVE_DISPLAY_AUTO should compatible with any type

this will introduce following errors

[       OK ] VaapiDisplayTest.CompatibleX11 (5 ms)
[ RUN      ] VaapiDisplayTest.CompatibleDRM
vaapidisplay_unittest.cpp:286: Failure
Value of: isCompatible(display, native)
  Actual: false
Expected: true
[  FAILED  ] VaapiDisplayTest.CompatibleDRM (1 ms)
[ RUN      ] VaapiDisplayTest.CompatibleVA
vaapidisplay_unittest.cpp:326: Failure
Value of: isCompatible(display, native)
  Actual: false
Expected: true
[  FAILED  ] VaapiDisplayTest.CompatibleVA (0 ms)
[----------] 7 tests from VaapiDisplayTest (15 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (15 ms total)
[  PASSED  ] 5 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] VaapiDisplayTest.CompatibleDRM
[  FAILED  ] VaapiDisplayTest.CompatibleVA

xuguangxin commented:
@uartie , Please help to review if you have time.
thanks

uartie reviewed (commented):
Why are you removing these tests?  It is checking that yami does the right thing when we give it an invalid handle.

uartie commented on vaapi/vaapidisplay_unittest.cpp:
```diff
@@ -101,19 +101,6 @@ class VaapiDisplayTest : public ::testing::Test {
-    EXPECT_FALSE(display.get());
-}
-#endif
-
```
Don't remove the test!  Fix VaapiDisplay::create!

uartie replied:
Ah yes, I see.  There's no way to determine if a pointer address is valid or not.  In that case, we should turn this into a DEATH test, instead, since we "expect" it to crash.

uartie replied:
See "Death Tests" in the docs: https://github.com/google/googletest/blob/master/googletest/docs/V1_7_AdvancedGuide.md#death-tests

xuguangxin replied:
I think Death Tests used for internal state checking.  In our case, we will crash in vaGetDisplay. It should not apply to death test, right?

uartie replied:
It doesn't matter where it crashes.  The point is we are testing that invalid input to our library function produces the "expected" outcome... in this case, we expect it to crash to tell the library user that they are doing something very wrong.  And a test such as this not only exercises our "negative" code path, it documents it too.

xuguangxin replied:
Ok, I can change it.thanks for explain.

uartie commented on vaapi/vaapidisplay_unittest.cpp:
```diff
@@ -134,19 +121,6 @@ VAAPI_DISPLAY_TEST(CreateInvalidHandleDRM) {
-
-    EXPECT_FALSE(display.get());
-}
-
```
Don't remove test!

uartie replied:
This test passes for me

xuguangxin replied:
It's passed only because it's a special value, 0 and -1, if you send -2 or 1 to create it will failed. Do we really need this test?

uartie replied:
There is nothing wrong with testing "special" values, especially for the "negative" path.  No matter how silly it may seem, there will always be someone who tries to do it.  We don't always want to test only the "good" (or "happy") path.  You never know when someone will come along and "accidentally" mess up the internal logic that handles 0 and -1 cases.  This test will catch that if no one else does.

IMO, any test is better than "no" test, no matter how trivial or inane it might seem.

BTW, if I use -2 or 1, I get a valid DisplayPtr... that is probably not good.  Seems to me that our intention is that a valid DisplayPtr means it's okay to use it.  In these cases, it is not okay.  Seems to me that we should be checking vaDisplayIsValid during the create function... and write separate DEATH tests to verify the expectation of such.

xuguangxin replied:
Ok, thanks for detailed explain, it make sense to me. I will change it.

xuguangxin committed:
vaapi: unittest, fix CreateInvalidHandleX11

use X11 display value 1 to create VaaiDisplay will crash the test. We
expect it death.

xuguangxin committed:
vaapi: display, check NATIVE_DISPLAY_VA handle

use vaDisplayIsValid to check VADisplay, this will crash
unit test in CreateInvalidHandleVA

xuguangxin committed:
vaapi: unittest, fix CreateInvalidHandleVA

this will fix crash introduced by previous commit

xuguangxin commented:
@uartie , fixed, please help to review.
thanks

uartie commented on vaapi/vaapidisplay_unittest.cpp:
```diff
@@ -103,14 +103,9 @@ class VaapiDisplayTest : public ::testing::Test {
-
-    // FIXME: This causes segmentation fault.
-    // VaapiDisplay should be able to handle this gracefully
     NativeDisplay native = {1, NATIVE_DISPLAY_X11};
```
Add `::testing::FLAGS_gtest_death_test_style = "threadsafe";` here to eliminate the gtest warning mentioned in your commit message

xuguangxin replied:
Little confused about this, we did not create new thread in unit test or library. Why gtest warning us about multithread things

uartie replied:
Gtest uses fork for death tests by default which it claims is unsafe if we are using threads in the library.  For some reason, gtest is "unable to detect" whether we are using threads in the library so it warns us.

AFAICT, instructing gtest to just use a thread for death tests instead of fork is harmless and better.  That is what the above code statement does.

xuguangxin replied:
thanks for explain, updated.

uartie commented on vaapi/vaapidisplay_unittest.cpp:
```diff
@@ -136,15 +131,16 @@ VAAPI_DISPLAY_TEST(CreateInvalidHandleDRM) {
 VAAPI_DISPLAY_TEST(CreateInvalidHandleVA) {
     NativeDisplay native = {0, NATIVE_DISPLAY_VA};
-    DisplayPtr display = VaapiDisplay::create(native);
 
```
Add ::testing::FLAGS_gtest_death_test_style = "threadsafe"; here.

uartie commented:
LGTM... but needs verified.  Thanks.

xuguangxin commented:
thanks you for reivew, we will start an internal test next week.  I will merge this and your factory refact code after this.

wudping commented:
verify + 1

xuguangxin commented:
merged as da4d1a527576d4eeb57ddfd92972acdc7a714e09. thanks U.Arite for review
